src/client/dfuse/pil4dfs/int_dfs.c

```diff
@@ -0,0 +1,5415 @@
+static pthread_mutex_t lock_fd_dup2ed;
+
+/* store ! umask to apply on mode when creating file to honor system umask */
+static mode_t          mode_not_umask;
+
+static void
+finalize_dfs(void);
+static void
+update_cwd(void);
+
+struct dir_hdl {
+	d_list_t   entry;
+	dfs_obj_t *oh;
+	char       name[DFS_MAX_PATH];
+};
+
+static inline struct dir_hdl *
+hdl_obj(d_list_t *rlink)
+{
+	return container_of(rlink, struct dir_hdl, entry);
+}
+
+static bool
+key_cmp(struct d_hash_table *htable, d_list_t *rlink, const void *key, unsigned int ksize)
+{
+	struct dir_hdl *hdl = hdl_obj(rlink);
+
+	return (strncmp(hdl->name, (const char *)key, DFS_MAX_PATH) == 0);
+}
+
+static void
+rec_free(struct d_hash_table *htable, d_list_t *rlink)
+{
+	int             rc;
+	struct dir_hdl *hdl = hdl_obj(rlink);
+
+	rc = dfs_release(hdl->oh);
+	if (rc == ENOMEM)
+		rc = dfs_release(hdl->oh);
+	if (rc && bLog)
+		D_WARN("dfs_release() failed: %s", strerror(rc));
+	D_FREE(hdl);
+}
+
+static bool
+rec_decref(struct d_hash_table *htable, d_list_t *rlink)
+{
+	return true;
+}
+
+static uint32_t
+rec_hash(struct d_hash_table *htable, d_list_t *rlink)
+{
+	struct dir_hdl *hdl = hdl_obj(rlink);
+
+	return d_hash_string_u32(hdl->name, strnlen(hdl->name, DFS_MAX_PATH));
+}
+
+static d_hash_table_ops_t hdl_hash_ops = {.hop_key_cmp    = key_cmp,
+					  .hop_rec_decref = rec_decref,
+					  .hop_rec_free   = rec_free,
+					  .hop_rec_hash   = rec_hash};
+
+/**
+ * Look up a path in hash table to get the dfs_obj for a dir.
+ * If not found in hash table, call to dfs_lookup() to open the object on DAOS.
+ * If the object is corresponding to a dir, insert the object into hash table.
+ * Since many DFS APIs need parent dir DFS object as parameter, we use a hash
+ * table to cache parent objects for efficiency.
+ */
+static int
+lookup_insert_dir(int idx_dfs, const char *name, mode_t *mode, dfs_obj_t **obj)
+{
+	struct dir_hdl *hdl;
+	dfs_obj_t      *oh;
+	d_list_t       *rlink;
+	size_t          len;
+	int             rc;
+
+	if (obj == NULL) {
+		D_ERROR("Invalid pointer!");
+		return EFAULT;
+	}
+	*obj = NULL;
+	len = strnlen(name, DFS_MAX_PATH);
+	if (len >= DFS_MAX_PATH) {
+		D_ERROR("path is too long!");
+		return ENAMETOOLONG;
+	}
+	rlink = d_hash_rec_find(dfs_list[idx_dfs].dfs_dir_hash, name, len);
+	if (rlink != NULL) {
+		hdl = hdl_obj(rlink);
+		*obj = hdl->oh;
+		return 0;
+	}
+
+	rc = dfs_lookup(dfs_list[idx_dfs].dfs, name, O_RDWR, &oh, mode, NULL);
+	if (rc)
+		return rc;
+
+	if (mode && !S_ISDIR(*mode)) {
+		*obj = oh;
+		return 0;
+	}
+
+	D_ALLOC_PTR(hdl);
+	if (hdl == NULL) {
+		dfs_release(oh);
+		return ENOMEM;
+	}
+
+	/* "len + 1" is used to avoid compiling issue on Ubuntu. */
+	strncpy(hdl->name, name, len + 1);
+	hdl->oh = oh;
+
+	rc = d_hash_rec_insert(dfs_list[idx_dfs].dfs_dir_hash, hdl->name, len, &hdl->entry, true);
+	if (rc) {
+		D_ERROR("Failed to insert dir handle in hashtable.");
+		dfs_release(hdl->oh);
+		D_FREE(hdl);
+		return daos_der2errno(rc);
+	}
```

Comment:
Failing with an error like this is only valid if you hold a lock over the rec_find() and rec_insert() call, to support multiple threads the best option is to change this to find_insert() and handle both cases here.

Other areas of daos will have different hash table usage because their use-case is different and that code might not be correct here.

Reply:
We will always need to call dfs_lookup() before d_hash_rec_find_insert() then. Calling dfs_lookup() is quite expensive. I can use lock when calling d_hash_rec_find() and d_hash_rec_insert().